tldw_Server_API/app/core/Prototype_Workspaces/service.py

```
@@ -0,0 +1,453 @@
+            },
+        )
+        updated_workspace = await self._repo.update_workspace_state(
+            prototype_workspace_id,
+            canonical_snapshot_id=candidate_snapshot_id,
+            last_known_good_snapshot_id=candidate_snapshot_id,
+            canonical_preview_status="ready",
+            publish_validation_status="validated",
+        )
+        if not updated_workspace or updated_workspace.get("canonical_snapshot_id") != candidate_snapshot_id:
+            await self._preview_broker.revoke_preview_handle(preview_grant["preview_handle"])
+            raise RuntimeError("failed to persist canonical workspace update")
```

Ruff (0.15.11)

[warning] Avoid specifying long messages outside the exception class

(TRY003)